File: LaLeMa-iOS/Sources/ViewModels/CalendarViewModel.swift
import Foundation
import SwiftUI

final class CalendarViewModel: ObservableObject {
    @Published var currentYearMonth: Date = Date()
    @Published var recordedDates: Set<String> = []
    @Published var selectedDate: String?
    @Published var selectedRecords: [PoopRecord] = []
    @Published var showRecordForm: Bool = false
    @Published var showDetailDialog: Bool = false

    private let repository = PoopRepository.shared
    private let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()

    private var calendar: Calendar {
        Calendar.current
    }

    init() {
        loadMonth()
    }

    func loadMonth() {
        let year = calendar.component(.year, from: currentYearMonth)
        let month = calendar.component(.month, from: currentYearMonth)

        let startOfMonth = calendar.date(from: DateComponents(year: year, month: month, day: 1))!
        let endOfMonth = calendar.date(byAdding: DateComponents(month: 1, day: -1), to: startOfMonth)!

        let startString = dateFormatter.string(from: startOfMonth)
        let endString = dateFormatter.string(from: endOfMonth)

        let records = repository.getByDateRange(startDate: startString, endDate: endString)
        recordedDates = Set(records.map { $0.date })
    }

    func previousMonth() {
        if let newDate = calendar.date(byAdding: .month, value: -1, to: currentYearMonth) {
            currentYearMonth = newDate
            loadMonth()
        }
    }

    func nextMonth() {
        if let newDate = calendar.date(byAdding: .month, value: 1, to: currentYearMonth) {
            currentYearMonth = newDate
            loadMonth()
        }
    }

    func onDateClick(_ date: String) {
        let today = Date()
        let todayString = dateFormatter.string(from: today)
        let clickedDate = dateFormatter.date(from: date)!

        if recordedDates.contains(date) {
            selectedRecords = repository.getByDate(date)
            selectedDate = date
            showDetailDialog = true
        } else if let daysDiff = calendar.dateComponents([.day], from: clickedDate, to: today).day,
                  daysDiff >= 0 && daysDiff <= 6 {
            selectedDate = date
            showRecordForm = true
        }
    }

    func deleteRecord(_ record: PoopRecord) {
        if let id = record.id {
            repository.deleteRecord(id: id)
            loadMonth()
            showDetailDialog = false
        }
    }

    func makeUpRecord(
        timeHour: Int,
        timeMinute: Int,
        amount: PoopAmount,
        consistency: PoopConsistency,
        color: PoopColor,
        smell: PoopSmell,
        painLevel: Int,
        blood: Bool,
        mucus: Bool,
        notes: String
    ) {
        guard let date = selectedDate else { return }

        _ = repository.record(
            date: date,
            timeHour: timeHour,
            timeMinute: timeMinute,
            amount: amount,
            consistency: consistency,
            color: color,
            smell: smell,
            painLevel: painLevel,
            blood: blood,
            mucus: mucus,
            notes: notes
        )

        loadMonth()
        showRecordForm = false
        selectedDate = nil
    }

    func dismissDialog() {
        showDetailDialog = false
        showRecordForm = false
        selectedDate = nil
        selectedRecords = []
    }

    var yearMonthString: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy年M月"
        return formatter.string(from: currentYearMonth)
    }
}

File: app/src/main/java/com/lalema/app/ui/calendar/CalendarViewModel.kt
package com.lalema.app.ui.calendar

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.lalema.app.data.PoopRecord
import com.lalema.app.domain.PoopRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.YearMonth
import java.time.format.DateTimeFormatter
import javax.inject.Inject

data class CalendarUiState(
    val currentYearMonth: YearMonth = YearMonth.now(),
    val recordedDates: Set<String> = emptySet(),
    val selectedDate: String? = null,
    val showMakeupDialog: Boolean = false,
    val showDetailDialog: Boolean = false,
    val showRecordForm: Boolean = false,
    val selectedRecords: List<PoopRecord> = emptyList()
)

@HiltViewModel
class CalendarViewModel @Inject constructor(
    private val repository: PoopRepository
) : ViewModel() {

    private val dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")

    private val _uiState = MutableStateFlow(CalendarUiState())
    val uiState: StateFlow<CalendarUiState> = _uiState.asStateFlow()

    init {
        loadMonth(YearMonth.now())
    }

    fun loadMonth(yearMonth: YearMonth) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(currentYearMonth = yearMonth)
            refreshMonthData()
        }
    }

    fun previousMonth() {
        loadMonth(_uiState.value.currentYearMonth.minusMonths(1))
    }

    fun nextMonth() {
        loadMonth(_uiState.value.currentYearMonth.plusMonths(1))
    }

    fun onDateClick(date: String) {
        val today = LocalDate.now()
        val clickedDate = LocalDate.parse(date, dateFormatter)
        val isFuture = clickedDate.isAfter(today)

        if (isFuture) return

        viewModelScope.launch {
            val records = repository.getByDate(date)
            _uiState.value = _uiState.value.copy(
                selectedDate = date,
                selectedRecords = records,
                showDetailDialog = true
            )
        }
    }

    fun makeupRecord(
        timeHour: Int,
        timeMinute: Int,
        amount: String,
        consistency: String,
        color: String,
        smell: String,
        painLevel: Int,
        blood: Boolean,
        mucus: Boolean,
        notes: String
    ) {
        viewModelScope.launch {
            repository.record(
                date = _uiState.value.selectedDate!!,
                timeHour = timeHour,
                timeMinute = timeMinute,
                amount = amount,
                consistency = consistency,
                color = color,
                smell = smell,
                painLevel = painLevel,
                blood = blood,
                mucus = mucus,
                notes = notes
            )
            refreshMonthData()
            _uiState.value = _uiState.value.copy(
                showRecordForm = false,
                selectedDate = null
            )
        }
    }

    fun deleteRecord(id: Long) {
        viewModelScope.launch {
            repository.deleteRecord(id)
            refreshMonthData()
            val date = _uiState.value.selectedDate
            if (date != null) {
                val records = repository.getByDate(date)
                _uiState.value = _uiState.value.copy(selectedRecords = records)
            }
        }
    }

    fun dismissDialog() {
        _uiState.value = _uiState.value.copy(
            showMakeupDialog = false,
            showDetailDialog = false,
            showRecordForm = false,
            selectedDate = null,
            selectedRecords = emptyList()
        )
    }

    fun showRecordFormForDate(date: String) {
        _uiState.value = _uiState.value.copy(
            selectedDate = date,
            showDetailDialog = false,
            showRecordForm = true
        )
    }

    private suspend fun refreshMonthData() {
        val yearMonth = _uiState.value.currentYearMonth
        val startDate = yearMonth.atDay(1).format(dateFormatter)
        val endDate = yearMonth.atEndOfMonth().format(dateFormatter)
        val records = repository.getByDateRange(startDate, endDate)
        val dates = records.map { it.date }.toSet()
        _uiState.value = _uiState.value.copy(recordedDates = dates)
    }

}
